### PR TITLE
VDS-1748 Implement csv class maps for contact and organization

### PR DESCRIPTION
**Pull request by trueboroda**
Description: 

--

QA-test: 

Demo-test: 

Download artifact URL: https://vc3prerelease.blob.core.windows.net/packages/VirtoCommerce.CustomerExportImport_1.0.0-pr-11-8a367fe5.zip

### CODE REVIEW

File: tests/VirtoCommerce.CustomerExportImportModule.Tests/CsvClassMapsTest.cs [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+            var selectedDynamicProperties = new[] { "Sex" };
+            csvWriter.Configuration.RegisterClassMap(new ContactClassMap(selectedDynamicProperties));
+
+            csvWriter.WriteRecords(new[] { exportableContact });
+
+            sw.Flush();
+            sw.Close();
+
+            csvWriter.Dispose();
+            sw.Dispose();
```
Comment (asvishnyakov) [inline tests/VirtoCommerce.CustomerExportImportModule.Tests/CsvClassMapsTest.cs:75]: I think Dispose should use Flush and Close internally

**Comment by mvktsk** [comment]
Review task https://virtocommerce.atlassian.net/browse/VP-7416 has been created

File: src/VirtoCommerce.CustomerExportImportModule.Core/Models/ExportableContact.cs [hunk 2]
```diff
@@ -92,44 +95,43 @@ public ExportableContact FromModel(Contact contact, Organization organization, S
...
+            FullName = contact.FullName;
+            ContactOuterId = contact.OuterId;
+            OrganizationId = organization?.Id;
+            OrganizationOuterId = organization?.OuterId;
+            OrganizationName = organization?.Name;
+            AccountId = account?.Id;
+            StoreId = account?.StoreId;
+            StoreName = store?.Name;
+            AccountLogin = account?.UserName;
+            AccountEmail = account?.StoreId;
```
Comment (asvishnyakov) [inline src/VirtoCommerce.CustomerExportImportModule.Core/Models/ExportableContact.cs:110]: Account.Email

File: src/VirtoCommerce.CustomerExportImportModule.Core/Models/ExportableContact.cs [hunk 3]
```diff
@@ -92,44 +95,43 @@ public ExportableContact FromModel(Contact contact, Organization organization, S
...
+            ContactStatus = contact.Status;
+            AssociatedOrganizationIds = contact.AssociatedOrganizations.IsNullOrEmpty() ? null : string.Join(", ", contact.AssociatedOrganizations);
+            BirthDate = contact.BirthDate;
+            TimeZone = contact.TimeZone;
+            Phones = contact.Phones.IsNullOrEmpty() ? null : string.Join(", ", contact.Phones);
+            UserGroups = contact.Groups.IsNullOrEmpty() ? null : string.Join(", ", contact.Groups);
+            AddressType = address?.AddressType.ToString();
+            AddressFirstName = address?.FirstName;
+            AddressLastName = address?.LastName;
+            AddressCountry = address?.RegionName;
```
Comment (asvishnyakov) [inline src/VirtoCommerce.CustomerExportImportModule.Core/Models/ExportableContact.cs:123]: Country, not RegionName

File: src/VirtoCommerce.CustomerExportImportModule.Core/Models/ExportableOrganization.cs [hunk 4]
```diff
@@ -41,33 +42,31 @@ public ExportableOrganization FromModel(Organization organization, Organization
...
+            Id = organization.Id;
+            OrganizationName = organization.Name;
+            OrganizationOuterId = organization.OuterId;
+            ParentOrganizationName = parentOrganization?.Name;
+            ParentOrganizationId = organization.ParentId;
+            ParentOrganizationOuterId = parentOrganization?.OuterId;
+            AddressType = address?.AddressType.ToString();
+            AddressFirstName = address?.FirstName;
+            AddressLastName = address?.LastName;
+            AddressCountry = address?.RegionName;
```
Comment (asvishnyakov) [inline src/VirtoCommerce.CustomerExportImportModule.Core/Models/ExportableOrganization.cs:54]: Country, not RegionName

**Review by asvishnyakov — CHANGES_REQUESTED** [review]

File: src/VirtoCommerce.CustomerExportImportModule.Data/ExportImport/GenericClassMap.cs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+                // Exporting multiple csv fields from the same property (which is a collection)
+                foreach (var dynamicProperty in dynamicProperties)
+                {
+                    // create CsvPropertyMap manually, because this.Map(x =>...) does not allow
+                    // to export multiple entries for the same property
+                    var propertyValuesInfo = exportedType.GetProperty(nameof(IHasDynamicProperties.DynamicProperties));
+                    var csvPropertyMap = MemberMap.CreateGeneric(exportedType, propertyValuesInfo);
+                    csvPropertyMap.Name(dynamicProperty);
+
+                    csvPropertyMap.Data.Index = ++columnIndex;
```
Comment (asvishnyakov) [inline src/VirtoCommerce.CustomerExportImportModule.Data/ExportImport/GenericClassMap.cs:35]: I'll prefer to write `var columnIndex = MemberMaps.Count` and `csvPropertyMap.DataIndex = columnIndex++`

File: src/VirtoCommerce.CustomerExportImportModule.Data/VirtoCommerce.CustomerExportImportModule.Data.csproj [hunk 6]
```diff
@@ -28,6 +27,7 @@
           <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
         </PackageReference>
         <PackageReference Include="Microsoft.SourceLink.GitHub" Version="1.0.0" PrivateAssets="All" />
+        <PackageReference Include="VirtoCommerce.CustomerModule.Core" Version="3.28.0" />
```
Comment (asvishnyakov) [inline src/VirtoCommerce.CustomerExportImportModule.Data/VirtoCommerce.CustomerExportImportModule.Data.csproj:30]: Core already has reference, not this is redundant

**Review by asvishnyakov — CHANGES_REQUESTED** [review]

**Review by asvishnyakov — CHANGES_REQUESTED** [review]
Please add tests for dictionary & multivalue dynamic properties, address mapping, and joining on enumerables.

**Review by asvishnyakov — APPROVED** [review]

**Pull request merged**